File: 2020/day/2/Main.hs
import Data.Char
import Data.List
import Text.ParserCombinators.ReadP

data Policy = Policy Int Int Char deriving Show

parsePolicy = Policy <$> number <* char '-' <*> number <* char ' ' <*> get
  where 
    number = read <$> munch1 isDigit

parsePolicyAndPassword = (,) <$> parsePolicy <*> (string ": " *> munch1 (/= '\n'))

valid1 (Policy lo hi x, password) = 
    case find ((== x) . head) (group (sort password)) of
        Nothing -> False
        Just ys -> lo <= length ys && hi >= length ys

valid2 (Policy lo hi x, password) = (password !! (lo - 1) == x) /= (password !! (hi - 1) == x)

main = do
    input <- readFile "input.txt"
    let [(policies, "")] = readP_to_S (sepBy1 parsePolicyAndPassword (char '\n') <* skipSpaces <* eof) input

    -- Part 1
    print $ length $ filter valid1 policies

    -- Part 2
    print $ length $ filter valid2 policies

File: 2020/day/2/main.py
import fileinput
from collections import Counter
import re

password_re = re.compile('^(\d+)-(\d+) (.): (.*)$')
ans1 = 0
ans2 = 0

with fileinput.input('input.txt') as f:
    for s in f:
        m = password_re.match(s.strip())

        # lo-hi x: password
        lo = int(m.group(1))
        hi = int(m.group(2))
        x = m.group(3)
        password = m.group(4)

        # Part 1
        if lo <= Counter(password).get(x, 0) <= hi:
            ans1 += 1

        # Part 2
        if (password[lo - 1] == x) != (password[hi - 1] == x):
            ans2 += 1

print(ans1)
print(ans2)